### ISSUE TITLE
Jep can sometimes pick the wrong method Java method to call

### ISSUE THREAD

**Issue by ndjensen** [ninia/jep#25]
Python doesn't have the concept of reusing the same method name with a different number of args like Java, if you want that you just use default args or keyword args.  This means a single call in a Jep interpreter to a Java method could call a number of methods.  A very good example within a single class is available on issue #11 with System.out.println(arg).

Tracing Jep's internals:
1. Jep looks at all methods on the object, including methods inherited from super classes
2. Jep filters the possible methods based on the Java method name matching the Python method name
3. Jep filters the possible methods based on the number of arguments matching
4. Jep loops over the remaining methods, comparing the argument types passed in from Python to the argument types on the Java method signature
5. As it checks each argument, it calls isAssignableFrom(pythonArgType, javaArgType)
6. If all of the arguments pass that check, it stops checking methods and calls that one
#11 is a problem because Jep is eagerly taking the first matching method instead of finding the _best_ matching method.

In theory this can get worse with inheritance though.  The [Java documentation for Class.getMethods()](http://docs.oracle.com/javase/7/docs/api/java/lang/Class.html#getMethods%28%29) states that

>  The elements in the array returned are not sorted and are not in any particular order.

`Class.getMethods()` will return the public methods, including the overridden super methods.  As an example, if by chance you had a `public class SomeList extends AbstractList<SomeType>`, there's a possibility it will return `AbstractList.add(Object)` ahead of `SomeList.add(SomeType)` in the method list.  When that occurs, since SomeType passes the isAssignableFrom check of being an Object, the super method would get called instead of the subclass' overridden method.

In practice most JVMs seem to be consistently returning a sane order of methods that has the subclass methods first.  At one point Sun/Oracle JDK 7 was not, and they refused to fix it.  http://bugs.java.com/bugdatabase/view_bug.do?bug_id=7023180  However, at some point after that the JVM seemed to return to a sane order again, dodging the problem with inheritance for a while longer.

Regardless, we need to improve method matching for #11 and the further scenario of inheritance, ensuring the subclass methods are picked over superclass methods where appropriate.

**Comment by ndjensen** [comment]
At present I've dreamt up a few solutions for this.  I know @bsteffensmeier thought up some other solutions such as a way to indicate to Jep how you want to handle this for your application....  My simple example is a Java class such as

```java
public interface JMath {

    public void doMath(double d);

    public void doMath(float f);

    public void doMath(Object o);

    public void doMath(List data);
}
```

My current thoughts on solutions:

1. Check all methods with matching names and number of args.  Implement algorithm that picks best match when there's more than one match.  For example, when evaluating the quality of the match, we could return an int value rating the quality of the match, and after evaluating all matches, go with the Java method that most closely matches.
  1.  When dealing with inheritance, we'd aim for methods with lower subclasses as args (e.g. Object as an arg returns a lesser quality value than anything else).  The above example would give the method with List a higher match value if List was compatible with the arg.
  2.  Subclass methods would have a higher match value over super class methods with the exact same signature.
  3.  When dealing with primitives, matching bit length would take precedence (e.g. double has a higher match value than float since python floats are 64-bit).  The above example would give the method with double a higher match value.
2. If numpy is installed, developers could use numpy dtypes to explicitly pick the type the arg should match in Java, in which case when dealing with primitives we'd always know which one to pick.  If the developer did a numpy.float32(value) or numpy.float64(value) we could know exactly which one to match.
3. A combination of the above enabling best match by default but allowing for explicit selection if necessary and possible.

**Comment by bsteffensmeier** [comment]
1. Option 1 is necessary, jep needs to have a deterministic way of picking methods.
 1. One problem I see is that when you implement multiple interfaces it might not be possible to pick a "lower" one.
 2. The second problem is that there might really be cases where you want to call a different method than jep picks, for example it would be impossible to call the float version of any method with a double version
2. Option 2 is a nice addition to option 1 it resolves the second problem I identifed, but only for primitive types and only if you use numpy.
3. So I like option 3, I think the two solutions build on eachother nicely but I still feel its incomplete.
4. If we want to be complete we need someway for people to access every method. From a design standpoint its just a matter of determining what sort of hoops you have to jump through to get that.
 1. Already today if you go through the java reflection API in your python you can eventually invoke any method on any class, if these cases are infrequent enough maybe that is enough support. We could make this solution a little easier if we automatically converted java.lang.reflect.Method into a pyjmethod when it is returned from a method because then it could easily be made callable which would feel more pythonic.
 2.  We could add a method to pyjobject to find the right method, taking classes as arguments. so JMath.resolve("doMath", Object)(arg).
 3. We have recently discussed making a pyjmultimethod, if we go that route it would make more sense to resolve methods on that object so the python ends up like: JMath.doMath.resolve(Object)(arg)
   1. That might make method binding harder, I was hoping to move away from pyjmethodwrapper and try to get the builtin python method to be used when we bind the callables but this would add custom functionality to bound methods so it would need to be in both the method wrapper and the multimethod.

Basically, I feel all the proposed solutions build on eachother, and we should implement any/all of them that we can.

**Comment by bsteffensmeier** [comment]
In fixing issue #42 I hopefully resolved a lot of the problems that people actually see. Any method that takes a single argument of a primitive type will be able to choose the correct overloaded method when passing in a python primitive. Some of the more complex cases are still not fixed but hopefully people are less likely to rely on these cases. The remaining issues include:

1. PyJObjects are not necessarily passed to the correct overloaded method. If the method takes an arg that is exactly the same class it will take priority but otherwise it is still non deterministic. So if you pass an ArrayList to an overloaded method that must choose between a method with a Collection arg and a method with a List arg then it will pick an arbitrary method. However if there is a method that takes an ArrayList then it will take priority. In theory the code would support better choosing easily by tying it into the same "scoring" system I am using for primitives, it is just a matter of writing the code to prioritize the methods.
2. We aren't making any strong guarantees if you overload a method with multiple args. For example if you pass in two python booleans to an overloaded method that can take (int, boolean) or (boolean,int) then its just going to pick one. There are many cases where it will be able to choose correctly, especially if you have primitive types or perfect matches but for multiple imperfect matches it is still a bit rough.
3. numpy primitive types still aren't supported
4. pyjmultimethod has been implemented but it still doesn't allow access to specific methods.

**Comment by bsteffensmeier** [comment]
In 949b01e3326d37bc12e8e8c3e44ef9d11b2d1df5 I have added more unit tests for the things that are deterministic and working correctly. I have also documented some of the things that don't work or are non deterministic. I intend to fix some of the easier cases for 3.7 and the unit tests will be updated to reflect that.

**Comment by bsteffensmeier** [comment]
In the 3.6 and 3.7 releases we have made substantial progress on this issue, all the known conflicts are resolved and the conversion code is consolidated and reasonably straightforward to adjust. I don't plan on spending any more time on this unless someone has a problem in real world use case so I am going to close this and anyone with problems can open a new issue about something more specific.
